Add support for mli files

jsomers opened the pull request:
Teach mdx to process the code blocks inside docstrings in mli files.

For instance, the following:

```ocaml
(** {[# List.map (fun x -> x * x) [1; 2; 3]]} *)
```

is corrected in-place to become:

```ocaml
(** {[# List.map (fun x -> x * x) [1; 2; 3]
      - : int list = [1; 4; 9]
    ]} *)
```

## Implementation

1. We lex mli files in two phases: first, we use the OCaml lexer (via `ppxlib`) to extract docstrings, then we use the odoc parser to extract code blocks within them. We then slice the input into contiguous regions of `Text` and `Block` items, just like the lexer for `Normal` syntax.

2. We add a new `Syntax.Mli` and thread `~syntax` arguments through a few more places so that we can properly format the output of mli's (avoiding newlines, etc.).

3. Since mli files are often written with formatters like ocamlformat and ocp-indent, and since these formatters might clobber the output of mdx, we make mdx's diff for the `Mli` syntax whitespace-invariant.

## Notes

* There is currently no support for header attributes (like `env=main`), because the odoc parser doesn't currently support an extended `{env=main[ let x = ... ]}` syntax for code blocks. This will have to be added later.

* We made changes to mdx to get the library to build within our environment (we're using Jenga). I also made the changes in this pull request against an earlier version of mdx, and wrote tests for them according to our internal conventions. I did my best to port the changes to this new version. But there are currently no tests and I'm not sure that this will build!

So I'd appreciate any instructions on how to get mdx running with dune (from scratch, more or less), and how to run the test suite.

NathanReb commented:
Thanks for working on this!

Regarding building and testing there's not much to know, `dune build` and `dune runtest` should do the trick. I'm happy to help you further with this dune if you have any further/more detailed questions.

The tests suite is plit between unit tests and end to end tests.

Unit tests live in [`test/lib/`](https://github.com/realworldocaml/mdx/tree/master/test/lib). We use [aclotest](https://github.com/mirage/alcotest) there. If you want to write unit tests you can take inspiration from [`test_block.ml`](https://github.com/realworldocaml/mdx/blob/master/test/lib/test_block.ml) for instance.

The end to end tests uses dune rules to call `mdx` subcommands and compare their output with reference files. They live in [`test/bin/`](https://github.com/realworldocaml/mdx/tree/master/test/bin).
You'll be interested in the [`mdx-test`](https://github.com/realworldocaml/mdx/tree/master/test/bin/mdx-test) subfolder since that's where the test for `ocaml-mdx test` are.
It's divided in three folders: `expect`, `failure` and `misc`. Each of those folder in turn contains one folder per test case. The `expect` one contains tests that ensure that the output of `ocaml-mdx test` on the input markdown file, using the options specified in `test-case.opts` is equal to `test-case.expected` (or to the input if there's no .expected file).
The `failure` is very similar except that it is used to test cases where `ocaml-mdx` will exit with a non-zero status.
Both of these folders dune rules are generated. If you wish to write a test case that don't fit any of these scenarios you can do it in the `misc` folder.

It's likely that you'll have to edit the rule generation a bit so that it works with `.mli` files as well. The shared parts of the rule generation can be found in this [`test/bin/gen_rule_helpers/`](https://github.com/realworldocaml/mdx/tree/master/test/bin/gen_rule_helpers).

I've been planning on adding a CONTRIBUTING.md to document the end to end test, I'll work on that ASAP!

jsomers commented:
Thanks! I think I've nearly got it building and running tests. I went through [many steps](https://gist.github.com/jsomers/a329588e1d23ba377e6b4ddb516996e3), and ultimately the tests are still exiting with status 1:
```sh
$ dune runtest
   ocaml-mdx test/bin/mdx-test/expect/errors.actual (exit 1)
(cd _build/default/test/bin/mdx-test/expect/errors && ../../../../../../install/default/bin/ocaml-mdx test --output - test-case.md) > _build/default/test/bin/mdx-test/expect/errors.actual
```

Any idea why that might be?

NathanReb commented:
Oh sorry, should have told you about setting up the dependencies. For future reference, once you have a working opam switch, you can simply run `opam install -t --deps-only .` from the root of the repo and it will install both regular and tests dependencies!

I'll add that to the README!

Also note that it should work with `lwt.5.x`, you might have been confused by lwt's warning!

I'm a bit confused by the tests errors but I'll give it a look!

NathanReb commented:
Ok so I dug into it a bit and it seems one of the new dependencies you added sets `Printexc.record_backtrace true` and it breaks the tests.

I suspect it comes from `ppxlib` (maybe indirectly) as I remember having to disable it myself there a few month ago, see https://github.com/ocaml-ppx/ppxlib/commit/622e1a30f0f0edddfe93879df78cd6c1ac671a6c#diff-8b1ef87e8b49371d4138edec195db0e5.

We probably want to disable them in mdx as well so my current work around was to do so in ``test/bin/main.ml` but I think we'd rather avoid the ppxlib dependency as it currently makes it hard to stay compatible with new versions of the compiler and requires a lot of maintenance to do so.

I'd suggest relying on minimal parser for mli files that doesn't depend on the AST. How doable do you think that is?

NathanReb commented:
In the meantime, I pushed the small fixes I had to make to get it to build and pass the tests to my fork, see: https://github.com/NathanReb/mdx/commit/a956cb53b529145bf3c261ca7be073bc02a1af2d

jsomers commented:
Thanks for those fixes, I've incorporated them into my branch (now pushed here). What's weird is that I had seen the test failures on a commit *before* my changes; I wonder if it has something to do with my environment.

Now, after pulling in your changes (and reintroducing the `ppxlib` dependency) I seem to be doing okay, except that I still get an `(exit 1)` on the tests:

```sh
$ dune runtest
   ocaml-mdx test/bin/mdx-test/expect/errors.actual (exit 1)
(cd _build/default/test/bin/mdx-test/expect/errors && ../../../../../../install/default/bin/ocaml-mdx test --output - test-case.md) > _build/default/test/bin/mdx-test/expect/errors.actual
```

There doesn't seem to be anything in that `errors.actual` file.

> I'd suggest relying on minimal parser for mli files that doesn't depend on the AST. How doable do you think that is?

I'll give that a try. Maybe I can get rid of the ppxlib AST dependency. I'll also try to fix these merge conflicts...

Thanks again!

jsomers commented:
The current version doesn't use `Ppxlib_ast` anymore, but rather `Ppxlib`. I've also fixed the merge conflicts. The build and tests seem to run okay, though still with that "exit 1" output I don't understand.

jsomers commented:
Okay, I removed the `ppxlib` dependency. There was no need for it---we can use OCaml's `Lexer` and `Parser` modules directly. This requires renaming `lexer.mll` because otherwise the OCaml `Lexer` module is shadowed by mdx's own lexer.

Locally at least `dune build` and `dune runtest` now run without error.

jsomers commented:
Okay, last comment from me: I added an end-to-end test along the lines you suggested. Do you think this is now ready for review?

NathanReb commented:
Yeah it should be! I'll take a look today!

Can you add the odoc dependency to the opam file so that the CI passes?

NathanReb commented:
Hmm could you also rebase on top of `master` to ease the review?

I think github isn't suited for merge-based workflows and shows a lot of unrelated changes in the diff.

jsomers commented:
@NathanReb I rebased against master (and squashed all the commits together) to make this easier to read. I also fixed some of the builds I think...

I'll try to add a few more tests.

Thanks @gpetiot for doing the rebase on that separate pull request. Would you mind closing that one now so that there's only one?

NathanReb commented on lib/mdx.ml:
```diff
@@ -73,31 +78,103 @@ let parse l =
 
+let remove_whitespace str =
+  let open Astring in
+  List.fold_right (fun s acc ->
```
This can be probably be simplified using `Astring`'s `String.filter Char.is_white`.

Could you also move it to `Util` please? That applies to other such functions you might have defined. The point is to avoid accidentally duplicate their code elsewhere if we ever need them.

jsomers replied:
Sounds good, done! (Will push shortly.)

NathanReb commented on lib/mdx.ml:
```diff
@@ -73,31 +78,103 @@ let parse l =
+  *)
+  let has_changed = match syntax with
+    | Syntax.Cram | Syntax.Normal -> corrected <> file_contents
+    | Syntax.Mli -> remove_whitespace corrected <> remove_whitespace file_contents
```
I'm a bit worried about removing all whitespaces for the diff here. It should work now because we only support toplevel blocks but I feel like it's easy to forget about this and break mli support if we had support for other blocks in mli files.

We don't necessarily have to fix this now but if we merge it as is we must open an issue as a reminder that we need to imrove that.

In particular, what I have in mind is tied to a refactoring of the tests logic where we would better split I/O from the generation of the updated blocks. Once we have that we can easily diff individual blocks and keep track of any a change there instead of diffing the whole document as a string. cc @gpetiot @voodoos.

jsomers replied:
This is important enough to our use case here that I think I will aim to leave it in -- and then address later along the lines you suggest.

NathanReb replied:
Sounds good, I'll add an issue once we merge!

NathanReb commented on lib/lexer_mdx.mll:
```diff
@@ -85,5 +85,6 @@ let token syntax lexbuf =
     match syntax with
     | Syntax.Normal -> text      None lexbuf
     | Syntax.Cram   -> cram_text None lexbuf
+    | Syntax.Mli    -> failwith "No lexer here for mli"
```
What do you think about splitting this into `markdown_token` and `cram_token` so that we get rid of the failwith and properly dispatch between the various parsers in `Mdx` where we have all we need to parse every syntax?

jsomers replied:
I'm not sure I know what you mean. Could you maybe sketch out what you're thinking this code would look like?

NathanReb replied:
What would you think if instead of having a:
```ocaml
val token : Syntax.t -> Lexing.lexbuf -> [...] list
```
we had:
```ocaml
val cram_token : Lexing.lexbuf -> [...] list
val markdown_token : Lexing.lexbuf -> [...] list
```
in `Lexer_mdx` and turned `Mdx.parse_lexbuf` into:
```ocaml
let parse_lexbuf file_contents syntax lexbuf =
  match syntax with
  | Mli -> Mli_parser.parse_mli file_contents
  | Normal -> parse (Lexer_mdx.markdown_token lexbuf)
  | Cram -> parse (Lexer_mdx.cram_token lexbuf)
```

So that we get rid of this unhandled `Mli` syntax in `Lexer_mdx`. We can clean it up further later but I'd rather get rid of that failwith since this `token` function is only used in one place, in which syntax is already known.

jsomers replied:
@NathanReb Ah, that's a good idea -- this is what we do now.

NathanReb commented on mdx.opam:
```diff
@@ -14,7 +14,7 @@ build: [
 
 depends: [
-  "ocaml" {>= "4.02.3"}
+  "ocaml" {>= "4.02.3" & < "4.10.0"}
```
Can you remove the `< 4.10.0` bound? There's nothing in mdx preventing it from working with 4.10. As soon as odoc gets a 4.10 compatible release we will be able to build there as well!

jsomers replied:
True! Done

NathanReb commented on test/bin/mdx-test/expect/simple-mli/test-case.mli:
```diff
@@ -0,0 +1,9 @@
+(** This is a doc comment with some code blocks in it:
+
+    {[# List.map (fun x -> x * x) [(1 + 9); 2; 3]
+- : int list = [100; 4; 9]
```
Looking at the test I'm wondering if it would hurt to have mdx respect the indentation or at least just indent everything to the block start level so that mdx users that don't use ocamlformat or don't have ocp-indent checked in CI can still reasonably use that feature.

That's still probably a lot of people so I'm kind of considering it. How much work do you think that would require?

jsomers replied:
I've done this because I agree it's better for those users. It wasn't a huge amount of work, just a lot of fiddly changes to get it right.

NathanReb commented on lib/mdx.ml:
```diff
@@ -73,31 +78,103 @@ let parse l =
       | `Block b   -> Block b
     ) l
 
+let parse_mli file_contents =
```
This could probably go `Mli_parser`.

It would be nice to have some unit tests for this btw, maybe for an intermediate step since `Block.t` are a bit heavyweight.

jsomers replied:
:+1: Moved it out and wrote some unit tests

jsomers commented on bin/test/main.ml:
```diff
@@ -393,6 +393,7 @@ let report_error_in_block block msg =
 let run setup non_deterministic not_verbose syntax silent verbose_findlib
     prelude prelude_str file section root direction force_output output : int =
     try
+      Printexc.record_backtrace false;
```
It looks like we do still need it, sadly -- at least the test/bin/mdx-test/expect/prelude-file/test-case.md test fails, printing this extra bit:

> Called from file "toplevel/toploop.ml", line 208, characters 17-27

NathanReb replied:
Ok, in that case I think it would be nice to disable backtraces by default but to expose a cli flag so that it can be controlled by users if they want to.

What do you think?

jsomers replied:
Sounds good, I will add the flag!

jsomers commented:
@NathanReb Besides the one open question [above](https://github.com/realworldocaml/mdx/pull/206#discussion_r373564367), this should be ready for your re-review. Thanks again!

NathanReb commented on .ocamlformat:
```diff
@@ -1,2 +1,2 @@
-version=0.13.0
```
This should still be `0.13.0`!

Sorry I added this comment to my previous review but it was still pending somehow :/

Julow commented on lib/mli_parser.ml:
```diff
@@ -0,0 +1,179 @@
+       The code below handles this multiline case, concatenating the included substrings
+       from lines 2-6 ([lines_to_include]). *)
+      let first_line = List.hd list in
+      let last_line = List.rev list |> List.hd in
```
This can be computed in a safe way by doing this before the match on `lines_to_include`:

```ocaml
let drop_last lst =
  match List.rev lst with
  | [] -> None
  | last :: rev_tl -> Some (List.rev rev_tl, last)

let drop_first_and_last = function
  | [] -> None
  | first :: tl -> Some (first, drop_last tl)
```

```ocaml
  match drop_first_and_last lines_to_include with
  | None-> ""
  | Some (line, None) -> String.sub line start.column (end_.column - start.column)
  | Some (first_line, Some (stripped, last_line)) ->
```

jsomers replied:
Quite right -- I've taken your suggestion, which is much better.

Julow commented on lib/mli_parser.ml:
```diff
@@ -0,0 +1,179 @@
+    String.split_on_char '\n' file_contents
+    |> List.mapi (fun i line -> (succ i, line))
+    |> List.filter (fun (lnum, _) -> start.line <= lnum && lnum <= end_.line)
+    |> List.map snd
```
This has a running time of `number of lines * number of comments`, that's `O(N^2)`.

What do you think of splitting the file into an array in `parse_mli` and accessing lines by their line number in the array ?
Another benefit of this is the computation of `end` in `parse_mli` will be far easier and more robust.

jsomers replied:
@Julow good call, I've done this.

jsomers commented on bin/cli.ml:
```diff
@@ -48,6 +48,12 @@ let silent_eval =
+  let doc = "Print backtraces when evaluating toplevel phrases." in
+  named
+    (fun x -> `Record_backtrace x)
+    Arg.(value & flag & info [ "record-backtrace" ] ~doc)
```
For context see https://github.com/realworldocaml/mdx/pull/206#issuecomment-575582487 and https://github.com/realworldocaml/mdx/pull/206#discussion_r373542330.

If there is another way to repair the broken test, I am happy to extract this into another pull request. But if I revert these changes I still get the error.

NathanReb replied:
@Julow I specifically asked for this change. Because one of our new dependencies has the side effect of enabling backtraces, we have to explicitly disable them in the code.

I want this to be configurable, even if the default is to disable backtraces.

Julow replied:
> I want this to be configurable, even if the default is to disable backtraces.

Why is disabling them in code not enough ?

NathanReb replied:
Because in some cases you might want to be able to show the backtraces in a toplevel code fragment.

I think it makes sense to add an escape hatch rather than turning them off with no proper way to re-enable them.

Backtraces being off is a good default, I'm just thinking about the other cases here.

Julow replied:
I think this shouldn't be included in this PR. My idea is that we should keep the previous behavior (backtraces disabled ?) and open an issue to talk about how to handle them correctly.

jsomers commented on lib/output.ml:
```diff
@@ -23,8 +23,11 @@ let dump ppf = function
+  | `Output s -> (
+      match syntax with
+      | Some Syntax.Mli -> Fmt.pf ppf "%a%s\n" pp_pad pad s
+      | _ -> Fmt.pf ppf "%a%s\n" pp_pad pad s )
```
Good call. This allows us to clean up some weaving-through of `?syntax` in a number of places.

jsomers commented on lib/mdx.ml:
```diff
@@ -74,10 +87,24 @@ type expect_result = Identical | Differs
+     ocamlformat, and it would be impossible to match this formatting exactly down to
+     every space and newline.
+
+     Instead, for this syntax only, we make the diff whitespace-invariant.
```
That's fine by me. I did this. I added a dependency on `str` but I assume that's okay, since it's part of the stdlib?

NathanReb replied:
That won't solve everything either as
```ocaml
# let x = "a  b"
```
will be considered equal to
```ocaml
# let x = "a    b"
```
If we want to be accurate we need to account for OCaml strings.

Julow replied:
Indeed we'll have a problem with code. Splitting words only works for text. Code should be compared verbatim.

NathanReb commented on bin/test/main.ml:
```diff
@@ -351,11 +359,14 @@ let report_error_in_block block msg =
-    run_exn setup non_deterministic silent_eval syntax silent verbose_findlib
-      prelude prelude_str file section root force_output output
+    ( match record_backtrace with
+    | `Record_backtrace bool -> Printexc.record_backtrace bool );
```
W should probably move that bit to `run_exn`!

Julow commented on lib/util.ml:
```diff
@@ -121,3 +123,11 @@ module List = struct
+module Array = struct
+  let slice t ~from ~to_ =
+    let start_index, length =
+      if from = to_ then (from, 1) else (from, to_ - from + 1)
```
The if is not necessary. If `from = to_`, `to_ - from + 1` is `1`.

Julow commented on lib/block.ml:
```diff
@@ -117,16 +120,38 @@ let dump ppf ({ file; line; section; labels; contents; value; _ } as b) =
+          s := Fmt.prefix (Fmt.unit " ") !s;
+          i := !i - 1
+        done;
+        !s
```
Is this the equivalent of `Fmt.fmt "%s%s" (String.make (t.column + 2) ' ')` ?

jsomers replied:
No :/. if I write:

```ocaml
let pp_lines syntax t =
  let pp =
    match syntax with
    | Some Syntax.Cram -> Fmt.fmt "  %s"
    | Some Syntax.Mli -> Fmt.fmt "%s%s" (String.make (t.column + 2) ' ')
    | _ -> Fmt.string
  in
  Fmt.(list ~sep:(unit "\n") pp)
```
I get the error:
```
File "lib/block.ml", line 131, characters 40-72:
Error: This expression has type string but an expression was expected of type
         Format.formatter
```

Julow replied:
Yes, sorry, the `ppf` argument comes first.
Also, padding with space characters is something `Format` does natively:
```ocaml
fun ppf -> Fmt.fmt "%*s%s" ppf (t.column + 2) ""
```

jsomers replied:
Apologies, but I declare defeat. I have never used the `Fmt` library before and find it hard to understand. I have struggled for many hours with `Fmt` inside mdx. Unless I'm given the exact diff to apply here that will compile, I won't be able to improve this particular bit of code.

Julow replied:
This diff works for me:
```dif
--- a/lib/block.ml
+++ b/lib/block.ml
@@ -128,14 +128,7 @@ let pp_lines syntax t =
   let pp =
     match syntax with
     | Some Syntax.Cram -> Fmt.fmt "  %s"
-    | Some Syntax.Mli ->
-        let i = ref (t.column + 2) in
-        let s = ref Fmt.string in
-        while !i > 0 do
-          s := Fmt.prefix (Fmt.unit " ") !s;
-          i := !i - 1
-        done;
-        !s
+    | Some Syntax.Mli -> fun ppf -> Fmt.fmt "%*s%s" ppf (t.column + 2) ""
     | _ -> Fmt.string
   in
   Fmt.(list ~sep:(unit "\n") pp)
```

jsomers replied:
what a dope i am! i get it now. see https://github.com/realworldocaml/mdx/pull/242

NathanReb commented:
> I found a loop between Mdx and OCamlformat
I thought OCaml format wasn't formatting the doc comments by default. Does that apply even with doc comment formatting off? If so it is indeed quite the issue, although I'm working on a dune feature to properly format files that are being promoted.

Do you confirm that ocamlformat won't mess with the content of the block and that if we get the rest right it should work fine?

Julow commented:
OCamlformat doesn't format docstrings by default for now, but that's a common used option that we consider turning on by default.

Maybe a fix to

jsomers commented:
> Indeed we'll have a problem with code. Splitting words only works for text. Code should be compared verbatim.

> I found a loop between Mdx and OCamlformat

@Julow, I wonder if we might kill two birds with one stone here by keeping the comparison verbatim, except being agnostic about boundary whitespace, i.e., the arrows in `{[<- ... ->]}`. I think that whitespace is the source of most of the problems, and it has little relevance to the code/output.

> Also giving this to Mdx again removes the closing of the comment:

```diff
 (** {[
       1 + 2 
-    ]} *)
+    ]}
```
This is a separate bug that I am currently investigating. Something weird is going on with the calculation of the "remainder" `Text` node at the end of the mli. In tests, strangely, if I make this portion long enough, it is added as a Text node, and if not, it is dropped. I'll pursue it more tomorrow.

Julow commented:
> @Julow, I wonder if we might kill two birds with one stone here by keeping the comparison verbatim, except being agnostic about boundary whitespace, i.e., the arrows in {[<- ... ->]}. I think that whitespace is the source of most of the problems, and it has little relevance to the code/output.

I think the comparison is not relevant, it is not used when `--force-output` is passed, which should be the typical use case with Dune.
The most important thing to do is to not change the input if possible (keeping the indentation and position of `{[`, `]}`). If this bigger problem is fixed, the comparison function can be much simpler.

jsomers commented:
@Julow I've fixed the bug where we eliminated the trailing `*)` and there should no longer be contention with ocamlformat on single-line blocks. Now, too, the comparison function is back to just being `String.equal` instead of splitting to words.

I don't think we should continue to block this pull request on fixing contention with external tooling, especially if that is not a default behavior of ocamlformat, and if there are changes coming to dune /mdx to integrate these formatters directly. The cost of fixing these issues inside this pull request is higher than fixing them later (e.g. because of keeping up with merge conflicts, and because it is harder to track the diffs here).

Julow commented on test/bin/mdx-test/expect/simple-mli/test-case.mli:
```diff
@@ -0,0 +1,20 @@
+      - : int list = [100; 4; 9]
+    ]}
+
+    {[List.map (fun x -> x * x) [1; 2; 3]]}
```
Could you add a multi line code block that is not toplevel ? Currently the `pp_lines` function is not tested.

jsomers replied:
Now that this has merged, I'll address this and your other pending comment in another pull request. Thanks!

jsomers replied:
https://github.com/realworldocaml/mdx/pull/242

NathanReb commented:
Wonderful! Thanks for working so hard this. Let's merge!
